src/pages/main/processor/hooks/useFileTreeEvent.ts

```diff
@@ -1,92 +1,171 @@
-import { useEffect } from "react";
+import { useCallback, useContext, useEffect, useRef } from "react";
 
 import { useAppState } from "@_redux/useAppState";
+import { useDispatch } from "react-redux";
+import {
+  FileTree_Event_RedoActionType,
+  FileTree_Event_UndoActionType,
+  TFileAction,
+  setDoingFileAction,
+  setFileAction,
+} from "@_redux/main/fileTree";
+import { callFileApi } from "@_node/apis";
+import { LogAllow } from "@_constants/global";
+import { TNodeUid } from "@_node/types";
+import { MainContext } from "@_redux/main";
+import { _path, confirmAlert, getFullnameFromUid } from "@_node/index";
+import { FileChangeAlertMessage } from "@_constants/main";
 
 export const useFileTreeEvent = () => {
-  const { fileAction, fileEventFutureLength } = useAppState();
+  const dispatch = useDispatch();
+  const {
+    project,
+    fileTree,
+    fileAction,
+    lastFileAction,
+    fileEventPastLength,
+    didUndo,
+    didRedo,
+  } = useAppState();
+  const {
+    fileHandlers,
+    addInvalidFileNodes,
+    removeInvalidFileNodes,
+    setReloadCurrentProjectTrigger,
+  } = useContext(MainContext);
 
-  // file event hms
+  const clearFutureHistoryTriggerRef = useRef(false);
+  const lastFileActionRef = useRef<TFileAction>({ action: null });
   useEffect(() => {
-    // reset fileAction in the new history
-    /* fileEventFutureLength === 0 &&
-      fileAction.type !== null &&
-      dispatch(setFileAction({ type: null })); */
-  }, [fileEventFutureLength]);
+    if (clearFutureHistoryTriggerRef.current) {
+      // remove invalid history events after `remove` action
```

Comment:
Are we performing file undo or redo when the active panel is `file`?

Reply:
yes, it is.